File: app/src/main/java/com/github/pockethub/android/ui/item/repository/RepositoryItem.java
package com.github.pockethub.android.ui.item.repository;

import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;

import com.github.pockethub.android.R;
import com.github.pockethub.android.ui.StyledText;
import com.github.pockethub.android.ui.item.BaseDataItem;
import com.github.pockethub.android.ui.item.BaseViewHolder;
import com.meisolsson.githubsdk.model.Repository;
import com.meisolsson.githubsdk.model.User;

import butterknife.BindView;

import static com.github.pockethub.android.ui.view.OcticonTextView.ICON_FORK;
import static com.github.pockethub.android.ui.view.OcticonTextView.ICON_MIRROR_PRIVATE;
import static com.github.pockethub.android.ui.view.OcticonTextView.ICON_MIRROR_PUBLIC;
import static com.github.pockethub.android.ui.view.OcticonTextView.ICON_PRIVATE;
import static com.github.pockethub.android.ui.view.OcticonTextView.ICON_PUBLIC;

public class RepositoryItem extends BaseDataItem<Repository, RepositoryItem.ViewHolder> {

    private final User user;
    private int descriptionColor = -1;

    public RepositoryItem(Repository dataItem, User user) {
        super(null, dataItem, dataItem.id());
        this.user = user;
    }

    @Override
    public void bind(@NonNull ViewHolder holder, int position) {
        Repository repo = getData();

        if (descriptionColor == -1) {
            descriptionColor = holder.getRoot().getResources().getColor(R.color.text_description);
        }

        StyledText name = new StyledText();
        if (user == null) {
            name.append(repo.owner().login()).append('/');
        } else {
            if (!user.login().equals(repo.owner().login())) {
                name.foreground(repo.owner().login(), descriptionColor)
                        .foreground('/', descriptionColor);
            }
        }

        name.bold(repo.name());
        holder.name.setText(name);

        if (TextUtils.isEmpty(repo.mirrorUrl())) {
            if (repo.isPrivate()) {
                holder.icon.setText(ICON_PRIVATE);
            } else if (repo.isFork()) {
                holder.icon.setText(ICON_FORK);
            } else {
                holder.icon.setText(ICON_PUBLIC);
            }
        } else {
            if (repo.isPrivate()) {
                holder.icon.setText(ICON_MIRROR_PRIVATE);
            } else {
                holder.icon.setText(ICON_MIRROR_PUBLIC);
            }
        }

        if (!TextUtils.isEmpty(repo.description())) {
            holder.description.setText(repo.description());
            holder.description.setVisibility(View.VISIBLE);
        } else {
            holder.description.setVisibility(View.GONE);
        }

        if (!TextUtils.isEmpty(repo.language())) {
            holder.language.setText(repo.language());
            holder.language.setVisibility(View.VISIBLE);
        } else {
            holder.language.setVisibility(View.GONE);
        }

        holder.watchers.setText(String.valueOf(repo.watchersCount()));
        holder.forks.setText(String.valueOf(repo.forksCount()));
    }

    @Override
    public int getLayout() {
        return R.layout.user_repo_item;
    }

    @Override
    public long getId() {
        return getData().id();
    }

    @NonNull
    @Override
    public ViewHolder createViewHolder(@NonNull View itemView) {
        return new ViewHolder(itemView);
    }

    class ViewHolder extends BaseViewHolder {

        @BindView(R.id.tv_repo_icon) TextView icon;
        @BindView(R.id.tv_repo_description) TextView description;
        @BindView(R.id.tv_language) TextView language;
        @BindView(R.id.tv_watchers) TextView watchers;
        @BindView(R.id.tv_forks) TextView forks;
        @BindView(R.id.tv_repo_name) TextView name;

        public ViewHolder(@NonNull View rootView) {
            super(rootView);
        }
    }
}

File: app/src/main/java/com/github/pockethub/android/ui/item/repository/RepositoryItem.kt
package com.github.pockethub.android.ui.item.repository

import android.text.TextUtils
import android.view.View
import androidx.core.text.bold
import androidx.core.text.buildSpannedString
import androidx.core.text.color
import com.github.pockethub.android.R
import com.github.pockethub.android.ui.view.OcticonTextView.ICON_FORK
import com.github.pockethub.android.ui.view.OcticonTextView.ICON_MIRROR_PRIVATE
import com.github.pockethub.android.ui.view.OcticonTextView.ICON_MIRROR_PUBLIC
import com.github.pockethub.android.ui.view.OcticonTextView.ICON_PRIVATE
import com.github.pockethub.android.ui.view.OcticonTextView.ICON_PUBLIC
import com.meisolsson.githubsdk.model.Repository
import com.meisolsson.githubsdk.model.User
import com.xwray.groupie.kotlinandroidextensions.Item
import com.xwray.groupie.kotlinandroidextensions.ViewHolder
import kotlinx.android.synthetic.main.repo_details.*
import kotlinx.android.synthetic.main.user_repo_item.*

class RepositoryItem(val repo: Repository, private val user: User?) : Item(repo.id()!!) {

    private var descriptionColor = -1

    override fun getId() = repo.id()!!

    override fun getLayout() = R.layout.user_repo_item

    override fun bind(holder: ViewHolder, position: Int) {
        if (descriptionColor == -1) {
            descriptionColor = holder.root.resources.getColor(R.color.text_description)
        }

        holder.tv_repo_name.text = buildSpannedString {
            if (user == null) {
                append(repo.owner()!!.login()).append('/')
            } else {
                if (user.login() != repo.owner()!!.login()) {
                    color(descriptionColor) {
                        append("${repo.owner()!!.login()}/")
                    }
                }
            }

            bold {
                append(repo.name())
            }
        }

        if (TextUtils.isEmpty(repo.mirrorUrl())) {
            when {
                repo.isPrivate!! -> holder.tv_repo_icon.text = ICON_PRIVATE
                repo.isFork!! -> holder.tv_repo_icon.text = ICON_FORK
                else -> holder.tv_repo_icon.text = ICON_PUBLIC
            }
        } else {
            if (repo.isPrivate!!) {
                holder.tv_repo_icon.text = ICON_MIRROR_PRIVATE
            } else {
                holder.tv_repo_icon.text = ICON_MIRROR_PUBLIC
            }
        }

        if (!TextUtils.isEmpty(repo.description())) {
            holder.tv_repo_description.text = repo.description()
            holder.tv_repo_description.visibility = View.VISIBLE
        } else {
            holder.tv_repo_description.visibility = View.GONE
        }

        if (!TextUtils.isEmpty(repo.language())) {
            holder.tv_language.text = repo.language()
            holder.tv_language.visibility = View.VISIBLE
        } else {
            holder.tv_language.visibility = View.GONE
        }

        holder.tv_watchers.text = repo.watchersCount().toString()
        holder.tv_forks.text = repo.forksCount().toString()
    }
}
